### PR TITLE
Add typechecking to the rest of expr

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: python/hail/expr/expression.py [hunk 1]
```diff
@@ -2045,37 +2159,37 @@ class NumericExpression(AtomicExpression):
...
-        ret_typ = self._bin_op_ret_typ(other)
-        return self._bin_op_reverse(name, other, ret_typ)
+    def _bin_op_numeric(self, name, other, f_override=None):
+        other = to_expr(other)
+        ret_typ, wrapper = self._bin_op_ret_typ(other)
+        if not ret_typ:
+            raise NotImplementedError("'{}' {} '{}'".format(
+                self._type, name, other._type))
+        if f_override:
+            ret_type = f_override(ret_typ)
```
Comment (jigold) [inline python/hail/expr/expression.py]: Fix to `ret_typ = ...`. This mistake is in a number of places!

File: python/hail/expr/types.py [hunk 2]
```diff
@@ -938,6 +938,12 @@ def _repr(self):
                    'is.hail.expr.types.TCallOptional$': (TCall, False),
                    'is.hail.expr.types.TCallRequired$': (TCall, True)}
 
+
+@typecheck(t=Type)
+def is_numeric(t):
```
Comment (jigold) [inline python/hail/expr/types.py]: Call this in expression.py.

File: python/hail/expr/expression.py [hunk 3]
```diff
@@ -1221,7 +1292,7 @@ def sort(self, ascending=True):
 
         Parameters
         ----------
-        ascending : bool or :py:class:`BooleanExpression`
+        ascending : :py:class:`BooleanExpression`
```
Comment (jigold) [inline python/hail/expr/expression.py]: There are a bunch of references that won't link. Missing ".". I propose we fix these in another PR.

File: python/hail/expr/expression.py [hunk 4]
```diff
@@ -605,7 +667,12 @@ def forall(self, f):
...
-        return self._bin_lambda_method("forall", f, self._type.element_type, lambda t: TBoolean())
+        def unify_ret(t):
+            if not isinstance(t, TBoolean):
+                raise TypeError("'forall' expects 'f' to return an expression of type 'Boolean', found '{}'".format(t))
+            return t
+
+        return self._bin_lambda_method("forall", f, self._type.element_type, unify_ret)
 
     def group_by(self, f):
         """Group elements into a dict according to a lambda function.
```
Comment (jigold) [inline python/hail/expr/expression.py]: Add a check that f is callable.

File: python/hail/expr/expression.py [hunk 5]
```diff
@@ -605,7 +667,12 @@ def forall(self, f):
         :py:class:`BooleanExpression`.
             ``True`` if `f` returns ``True`` for every element, ``False`` otherwise.
         """
-        return self._bin_lambda_method("forall", f, self._type.element_type, lambda t: TBoolean())
+        def unify_ret(t):
+            if not isinstance(t, TBoolean):
+                raise TypeError("'forall' expects 'f' to return an expression of type 'Boolean', found '{}'".format(t))
+            return t
+
+        return self._bin_lambda_method("forall", f, self._type.element_type, unify_ret)
```
Comment (jigold) [inline python/hail/expr/expression.py]: check f is callable. Same for other functions in this class.

File: python/hail/expr/expression.py [hunk 6]
```diff
@@ -250,6 +255,32 @@ def unify_all(*exprs):
...
+        # assert the two classes are different
+        assert len(s) == 2
+        if TFloat64 in s:
+            return TFloat64()
+        elif TFloat32 in s:
+            return TFloat32()
+        else:
+            assert s == {TInt32, TInt64}
+            return TInt64()
+    elif c1 == TArray and c2 == TArray:
```
Comment (jigold) [inline python/hail/expr/expression.py]: Should this handle the Set case as well?

File: python/hail/expr/aggregators.py [hunk 7]
```diff
@@ -623,11 +640,13 @@ def explode(expr):
         Aggregable expression.
     """
     agg = _to_agg(expr)
+    if not (isinstance(agg._type, TSet) or isinstance(agg._type, TArray)):
+        raise  TypeError("'explode' expects a 'Set' or 'Array' argument, found '{}'".format(agg._type))
     uid = Env._get_uid()
     return Aggregable(LambdaClassMethod('flatMap', uid, agg._ast, Reference(uid)),
-                      agg._type, agg._indices, agg._aggregations, agg._joins, agg._refs)
-
+                      agg._type.element_type, agg._indices, agg._aggregations, agg._joins, agg._refs)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/aggregators.py]: it's used in the tests for sure. we should flesh the tests out but should do it separately!

File: python/hail/expr/aggregators.py [hunk 8]
```diff
@@ -204,7 +205,7 @@ def counter(expr):
     agg = _to_agg(expr)
     return _agg_func('counter', agg, TDict(agg._type, TInt64()))
 
-
+@typecheck(expr=oneof(Aggregable, expr_any), n=integral, ordering=nullable(expr_any))
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/aggregators.py]: I think this is okay. these files are long enough!!!

File: python/hail/expr/expression.py (hunk 6 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: It looks like we don't handle this in expr right now:
```
<input>:1:if (true) [1.0, 2.0].toSet() else [3, 4].toSet()
          ^

Java stack trace:
is.hail.utils.HailException: expected same-type `then' and `else' clause, got `Set[Float64]' and `Set[Int32]'
<input>:1:if (true) [1.0, 2.0].toSet() else [3, 4].toSet()
```

File: python/hail/expr/expression.py (hunk 4 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: good 'call'! I'll add a function typechecker.

File: python/hail/expr/expression.py [hunk 9]
```diff
@@ -250,6 +255,32 @@ def unify_all(*exprs):
...
+        et = unify_types(t1.element_type, t2.element_type)
+        if et:
+            return TArray(et)
+        else:
+            return None
+    else:
+        return None
+
 __numeric_types = [TInt32, TInt64, TFloat32, TFloat64]
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/expression.py]: oh, yeah, totally.

File: python/hail/expr/expression.py (hunk 6 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: actually, this is a problem in ifs too:
```
<input>:1:if (true) [1.0, 2.0] else [3, 4]
          ^

Java stack trace:
is.hail.utils.HailException: expected same-type `then' and `else' clause, got `Array[Float64]' and `Array[Int32]'
<input>:1:if (true) [1.0, 2.0] else [3, 4]
```

Comment (tpoterba) [reply python/hail/expr/expression.py]: very good catch! will fix in the scala code for this PR.

Comment (tpoterba) [reply python/hail/expr/expression.py]: on second thought, delaying this until later. But am catching the error in python!

File: python/hail/expr/expression.py [hunk 10]
```diff
@@ -1011,62 +1093,34 @@ class ArrayNumericExpression(ArrayExpression, CollectionNumericExpression):
...
+        if not t:
+            return None
         else:
-            raise NotImplementedError('''Error in return type for numeric conversion.
-                self = {}
-                other = {}'''.format(type(self), type(other)))
+            return TArray(t)
 
-    def _bin_op_numeric(self, name, other):
+    def _bin_op_numeric(self, name, other, f_override=None):
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/expr/expression.py]: yep, good call.

File: python/hail/expr/expression.py (hunk 3 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: yeah, this is a problem everywhere.

File: python/hail/expr/aggregators.py [hunk 11]
```diff
@@ -670,7 +670,7 @@ def filter(condition, expr):
 
     Parameters
     ----------
-    condition : :class:`hail.expr.expression.BooleanExpression` or function
+    condition : :class:`hail.expr.expression.BooleanExpression` or 1-argument function
```
Comment (jigold) [inline python/hail/expr/aggregators.py]: Is there a way to concisely say the function return type must be boolean? I see you check the result is boolean below.

File: python/hail/expr/expression.py [hunk 12]
```diff
@@ -150,37 +146,26 @@ def to_expr(e):
...
-
         kc = to_expr(keys)
         vc = to_expr(values)
 
         indices, aggregations, joins, refs = unify_all(kc, vc)
 
-        assert key_type == kc._type.element_type
-        assert value_type == vc._type.element_type
+        assert kt == kc._type.element_type
+        assert vt== vc._type.element_type
```
Comment (jigold) [inline python/hail/expr/expression.py]: space after `vt`

File: python/hail/expr/aggregators.py (hunk 11 above)
Comment (tpoterba) [reply python/hail/expr/aggregators.py]: oooo, good idea. We can do `func_spec(1, expr_bool)`. Great idea!

**Pull request merged**